theblues/charmstore.py

```diff
@@ -284,6 +284,65 @@ def search(self, text, includes=None, doc_type=None, limit=None,
         data = self._get(url)
         return data.json()['Results']
 
+    def list(self, includes=None, doc_type=None, promulgated_only=False,
+             sort=None, owner=None, series=None):
+        '''
+        List entities in the charmstore.
+
+        @param includes What metadata to return in results (e.g. charm-config)
+        @param doc_type Filter to this type: bundle or charm
+        @param promulgated_only Whether to filter to only promulgated charms.
+        @param sort Sorting the result based on the sort string provided
+        which can be name, author, series and - in front for descending
+        @param owner Optional owner. If provided, search results will only
+        include entities that owner can view.
+        @param series The series to filter; can be a list of series or a
+        single series.
+        '''
+        url = '%s/list' % (self.url)
+        first = True
+        if includes is not None:
+            includes = '&'.join(['include=%s' % i for i in includes])
+            if first:
```

Comment:
Rather than using this manual query building (with all the `if first` checks) I'd use `urllib.urlencode`.
For instance:
```python
queries = []
if includes is not None:
    queries.extend([('include', include) for include in includes])
if doc_type is not None:
    queries.append(('type', doc_type))
if promulgated_only:
    queries.append(('promulgated', 1))
... # etc, for the remaining keys
url = '{}/list?{}'.format(url, urllib.urlencode(queries))
```